core/src/main/scala/io/projectglow/plink/PlinkRowToInternalRowConverter.scala

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2019 The Glow Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.projectglow.plink
+
+import org.apache.spark.sql.SQLUtils.structFieldsEqualExceptNullability
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.util.{ArrayData, GenericArrayData}
+import org.apache.spark.sql.types.{ArrayType, StructType}
+import org.apache.spark.unsafe.types.UTF8String
+
+import io.projectglow.common.{BgenGenotype, BgenRow, GlowLogging, VariantSchemas}
+import io.projectglow.sql.util.RowConverter
+
+/**
+ * Creates an [[InternalRow]] with a given required schema. During construction,
+ * this class will throw an [[IllegalArgumentException]] if any of the fields in the required
+ * schema cannot be derived from a PLINK record.
+ */
+class PlinkRowToInternalRowConverter(schema: StructType) extends GlowLogging {
+  import io.projectglow.common.VariantSchemas._
+  import BimRow._
+
+  private val converter = {
+    val fns = schema.map { field =>
+      val fn: RowConverter.Updater[(InternalRow, Array[String], Array[Array[Int]])] = field match {
+        case f if structFieldsEqualExceptNullability(f, contigNameField) =>
```

Comment:
It's kind of sketchy to write manual converters for these fields that are pulled from one internal row and put into another. We should be able to use one of Spark's codegen'd projections to convert from the parsed bim row into the output schema. Then we only need to update the genotypes manually.